### PR TITLE
[Gutenberg] - Add support for unselecting blocks with the hardware back button

### PR DESCRIPTION
**Pull request by geriux**
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6483
- https://github.com/WordPress/gutenberg/pull/57279

This PR adds the functionality to be able to intercept the hardware back button from the React Native app.

It adds a `OnBackPressedCallback` in `GutenbergContainerFragment` to be able to manage the back button events from the React Native's BackHandler API.

If the Gutenberg editor adds a new back button subscription it will first handle that and then it will bubble up the event attached in the `EditPostActivity`

-----

## To Test:

Follow the test case scenarios in the [Gutenberg PR description](https://github.com/WordPress/gutenberg/pull/57279).

-----

## Regression Notes

1. Potential unintended areas of impact

    - It should only affect the editor.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - Manual testing and existing UI tests.

3. What automated tests I added (or what prevented me from doing so)

    - TODO

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes Testing Checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergContainerFragment.java [hunk 1]
```diff
@@ -170,6 +174,14 @@ public void onPause() {
...
         super.onResume();
 
+        OnBackPressedCallback callback = new OnBackPressedCallback(true) {
+            @Override
+            public void handleOnBackPressed() {
+                mWPAndroidGlueCode.onBackPressed();
+            }
+        };
+
+        getActivity().getOnBackPressedDispatcher().addCallback(this, callback);
```
Comment (fluiddot) [inline libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergContainerFragment.java:192]: I thought we might need to remove the callback when the fragment is destroyed, however, seems this is already covered by Android ([reference](https://developer.android.com/reference/androidx/activity/OnBackPressedDispatcher#addCallback(androidx.lifecycle.LifecycleOwner,androidx.activity.OnBackPressedCallback))).

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -1518,6 +1518,10 @@ private boolean handleBackPressed() {
         return true;
     }
 
+    public void onEditorBackPressed() {
+        getOnBackPressedDispatcher().onBackPressed();
+    }
+
```
Comment (fluiddot) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I couldn't find a call to this method across the code base, so I wonder if we need it.

Comment (geriux) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Nice catch! This was the first approach I took 😅 

This is handled by `onBackHandlerButton`. I'll remove it, thanks!

Comment (geriux) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: Removed in https://github.com/wordpress-mobile/WordPress-Android/pull/19828/commits/96601500e2335cb70d0d79ff75bfdf78e507ce8b

File: libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergContainerFragment.java (hunk 1 above)
Comment (geriux) [reply libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergContainerFragment.java:192]: Yup I thought that too but by looking at other places in the app confirmed it wasn't needed as well as the docs reference you mentioned 😄

File: RELEASE-NOTES.txt [hunk 3]
```diff
@@ -5,6 +5,7 @@
 * [*] Filter media types when sharing files to the editor [https://github.com/wordpress-mobile/WordPress-Android/pull/19754]
 * [***] [Jetpack-only] Plans: Upgrade to a Plan from domains dashboard in Jetpack app [https://github.com/wordpress-mobile/WordPress-Android/pull/19818]
 * [**] [internal] Removes unused resources and code [https://github.com/wordpress-mobile/WordPress-Android/pull/19788]
+* [*] Add support for unselecting blocks with the hardware back button [https://github.com/wordpress-mobile/WordPress-Android/pull/19828]
```
Comment (fluiddot) [inline RELEASE-NOTES.txt]: Nitpick: I'd add the prefix `Block Editor:` to highlight that this is a change related to the editor.

**Review by fluiddot — APPROVED** [review]
Approved via https://github.com/WordPress/gutenberg/pull/57279#pullrequestreview-1800276649.

**Pull request merged**